### PR TITLE
Refactor `thrust::complex` as a struct derived from `cuda::std::complex`

### PR DESCRIPTION
**Pull request by Blonck**
This commit refactors the thrust::complex type to be a struct derived
from `cuda::std::complex`, enabling reuse of existing implementation
logic. However, to maintain backward compatibility, certain
operators are reintroduced to allow type promotion between `float`
and `double` for the underlying type.

This PR is derived from #260 and has been rebased onto the current main branch

## Checklist
- [x] I am familiar with the [Contributing Guidelines]().

### CODE REVIEW

File: thrust/thrust/complex.h
Comment (jrhemstad) [inline thrust/thrust/complex.h:91]: Could we avoid this by making `thrust::complex` an alias to `cuda::std::complex`? Or is there custom logic that we need to implement directly in the `thrust::complex` type that `cuda::std::complex` doesn't provide?

Comment (Blonck) [reply thrust/thrust/complex.h:91]: Unfortunately, the custom logic is needed for backwards compatibility.
This is also the reason the unit test for `thrust::complex` were extended in a separate PR (https://github.com/NVIDIA/cccl/pull/413) to test that this change is BC and also does not introduce additional functionality. Sorry, should have mentioned that in the PR description.

**Comment by miscco** [comment]
So this is failing because I removed the long double overloads that are actually removed and static asserted that we should never use them. I believe we should also remove that test

**Review by miscco — APPROVED** [review]
This is awesome 🎉 

Thanks a lot for putting all the work in and improving our complex implementation.

Given that I had some part in writing some of the code I would want a second maintainers eyes  before we merge though

File: libcudacxx/include/cuda/std/detail/libcxx/include/complex
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/complex:1191]: @griwes This was preexisting in libcu++ 

I consider this a preexisting bug. Do you have any objections in disabling `long double` when it is not defined?

File: thrust/thrust/complex.h
Comment (miscco) [inline thrust/thrust/complex.h:71]: Nitpick: This requires instantiation of two types. We could improve compile time by moving the enable if into the template arguments:
```suggestion
  template <class U, typename detail::enable_if<!detail::is_same<T, U>::value, int> = 0,
                     typename detail::enable_if<detail::has_promoted_numerical_type<T, U>, int> = 0>
  __host__ __device__
  complex &operator=(const complex<U> &rhs)
  {
    this->real(static_cast<T>(rhs.real()));
    this->imag(static_cast<T>(rhs.imag()));
    return *this;
  }
```

This will short circuit if the first condition is not satisfied

**Comment by miscco** [comment]
The complex files were under an old apache license we should clean the new files up given that we completely reimplemented it

**Pull request merged**
